Command-line control over scalar-valued set routines

drreynolds opened the pull request:
I believe this PR is now complete, although we should hold off on reviews until the current release is out.

Steven-Roberts commented on include/arkode/arkode.h:
```diff
@@ -214,6 +214,10 @@ typedef enum
  * -------------------------- */
 
+/* Command-line control over ARKODE options */
+SUNDIALS_EXPORT int ARKodeSetFromCommandLine(void* arkode_mem, int argc,
+                                             char* argv[]);
```
```suggestion
SUNDIALS_EXPORT int ARKodeSetFromCommandLine(void* arkode_mem, int argc,
                                             const char* argv[]);
```

Steven-Roberts commented on src/arkode/arkode_io.c:
```diff
@@ -34,6 +35,347 @@
+  ark_mem = (ARKodeMem)arkode_mem;
+
+  /* Set list of integer command-line arguments, and the corresponding set routine */
+  int num_int_keys = 15;
+  const char* int_keys[] = {"arkode.order",
+                            "arkode.interpolant_degree",
+                            "arkode.linear",
+                            "arkode.autonomous",
+                            "arkode.deduce_implicit_rhs",
+                            "arkode.lsetup_frequency",
+                            "arkode.predictor_method",
+                            "arkode.max_nonlin_iters",
+                            "arkode.max_hnil_warns",
+                            "arkode.interpolate_stop_time",
+                            "arkode.max_num_constr_fails",
+                            "arkode.adaptivity_adjustment",
+                            "arkode.small_num_efails",
+                            "arkode.max_err_test_fails",
+                            "arkode.max_conv_fails"};
+  arkIntSetFn int_set[] = {ARKodeSetOrder,
+                           ARKodeSetInterpolantDegree,
+                           ARKodeSetLinear,
+                           ARKodeSetAutonomous,
+                           ARKodeSetDeduceImplicitRhs,
+                           ARKodeSetLSetupFrequency,
+                           ARKodeSetPredictorMethod,
+                           ARKodeSetMaxNonlinIters,
+                           ARKodeSetMaxHnilWarns,
+                           ARKodeSetInterpolateStopTime,
+                           ARKodeSetMaxNumConstrFails,
+                           ARKodeSetAdaptivityAdjustment,
+                           ARKodeSetSmallNumEFails,
+                           ARKodeSetMaxErrTestFails,
+                           ARKodeSetMaxConvFails};
```
```suggestion
  static const char* int_keys[] = {"arkode.order",
                            "arkode.interpolant_degree",
                            "arkode.linear",
                            "arkode.autonomous",
                            "arkode.deduce_implicit_rhs",
                            "arkode.lsetup_frequency",
                            "arkode.predictor_method",
                            "arkode.max_nonlin_iters",
                            "arkode.max_hnil_warns",
                            "arkode.interpolate_stop_time",
                            "arkode.max_num_constr_fails",
                            "arkode.adaptivity_adjustment",
                            "arkode.small_num_efails",
                            "arkode.max_err_test_fails",
                            "arkode.max_conv_fails"};
  static arkIntSetFn int_set[] = {ARKodeSetOrder,
                           ARKodeSetInterpolantDegree,
                           ARKodeSetLinear,
                           ARKodeSetAutonomous,
                           ARKodeSetDeduceImplicitRhs,
                           ARKodeSetLSetupFrequency,
                           ARKodeSetPredictorMethod,
                           ARKodeSetMaxNonlinIters,
                           ARKodeSetMaxHnilWarns,
                           ARKodeSetInterpolateStopTime,
                           ARKodeSetMaxNumConstrFails,
                           ARKodeSetAdaptivityAdjustment,
                           ARKodeSetSmallNumEFails,
                           ARKodeSetMaxErrTestFails,
                           ARKodeSetMaxConvFails};
  static const int num_int_keys = sizeof(int_keys) / sizeof(*int_keys);
```
and similar changes for code below...
To avoid issues with maintaining consistent lengths of these arrays, a cleaner solution might be to have a single array of `struct { char* name; arkIntSetFn fun}`.
Alternatively, this seems like an ideal place to use SUNDIALS' hashmap.

gardner48 replied:
I agree using the hashmap would be a good way to make sure these lists to not get out of sync somehow.

Steven-Roberts replied:
Unlike the current arrays, a hashmap would not be stored fully as a static global variable since it uses malloc. We could declare it as a static local variable that only gets initialized once, but then I worry about thread safety.

Now I'm thinking arrays might be safer. If we want to keep fast lookup times, we can ensure they are sorted and use `bsearch` rather than manually iterating over them.

gardner48 replied:
Good point, maybe using an array of structs would be a better way to go to keep the `name` and `fun` together but not need to `malloc` the array.

drreynolds replied:
I like the plan for an array of structs.  I haven't seen bsearch before, so I'll need to read up a bit on that instead of just iterating through the list.

Steven-Roberts commented on src/arkode/arkode_io.c:
```diff
@@ -34,6 +35,347 @@
+  *arg_used = SUNFALSE;
+  if (strcmp(argv[*i], argtest) == 0)
+  {
+    int iarg = atoi(argv[++(*i)]);
```
This usage of `++` is confusing. Can it be done on a separate line?

Steven-Roberts commented on src/arkode/arkode_io.c:
```diff
@@ -34,6 +35,347 @@
+                              sunbooleantype *arg_used)
+{
+  *arg_used = SUNFALSE;
+  if (strcmp(argv[*i], argtest) == 0)
```
So booleans are specified with just the flag? I think it would be preferable to have flag-value pair command line syntax for boolean options. Otherwise, a user will need to know the default and whether to override it.

gardner48 replied:
I think it would be clearer to refer to these a "aciton" functions rather than boolean functions as their only input is the integrator memory structure e.g., `ARKodeSetNonlinear` and `ARKodeClearStopTime`. The functions that take a boolean input are group with the "int" functions.

Steven-Roberts commented on src/arkode/arkode_io.c:
```diff
@@ -34,6 +35,347 @@
+  arkBoolSetFn bool_set[] = {ARKodeSetNonlinear,
+                             ARKodeClearStopTime,
+                             ARKodeSetNoInactiveRootWarn,
+                             ARKodeResetAccumulatedError};
```
It seems there's two type of boolean set functions to consider. Ones like `ARKodeResetAccumulatedError` don't take a boolean as input while ones like `ARKodeSetAutonomous` do. Those might need to be handled separately in this parsing.

Steven-Roberts replied:
Nevermind, I overlooked that `ARKodeSetAutonomous` is already covered in `arkIntSetFn`

Steven-Roberts commented on src/arkode/arkode_io.c:
```diff
@@ -34,6 +35,347 @@
+    sunbooleantype arg_used = SUNFALSE;
+
+    /* skip command-line arguments that do not begin with "arkode." */
+    if (strncmp(argv[i], "arkode.", 7) != 0) { continue; }
```
Since we filter by the `arkode.` prefix here, I think it's redundant to include it in all the array entries above. We would just need to shift the string pointer below.

Steven-Roberts replied:
Possibly out of scope for this initial PR, but in order to support full configuration over an MRI or operator splitting method where they may be several arkode instances, we may need to support user-specified prefixes. For example,  `arkode.order 2 partition1.order 2 partition2.linear 1`

gardner48 replied:
I think we could do this by requiring a certain format for the keys e.g., `user_label.arkode.option` or `arkode.user_label.option`. Then `ARKodeSetFromCommandLine` could accept an additional input for which `user_label` to filter for and if none is provided, only read `arkode.` keys.

balos1 replied:
We have a rough sketch of scheme for MRI and operator splitting in the design document.

gardner48 replied:
Ah, I remember this now. Looks like we settled on a `pkg.scope.option` style.

If this functions is going to be at the  ARKODE level, I think it will need an additional input to filter with since this function won't know which MRI level or splitting partition the memory corresponds to. 

Will we also need stepper specific versions of this function to handle options like setting the method to use or should we have only stepper-specific versions that call this function internally?

Steven-Roberts commented on src/arkode/arkode_io.c:
```diff
@@ -34,6 +35,347 @@
+    sunbooleantype arg_used = SUNFALSE;
+
+    /* skip command-line arguments that do not begin with "arkode." */
+    if (strncmp(argv[i], "arkode.", 7) != 0) { continue; }
```
Removes "magic constant"
```suggestion
    static const char *prefix = "arkode.";
    if (strncmp(argv[i], prefix, strlen(prefix)) != 0) { continue; }
```

Steven-Roberts commented on src/arkode/arkode_io.c:
```diff
@@ -34,6 +35,347 @@
+    if (strcmp(argv[i], "arkode.interpolant_type") == 0)
+    {
+      i++;
+      retval = 1;
```
```suggestion
      retval = ARK_ILL_INPUT;
```

Steven-Roberts commented on src/arkode/arkode_io.c:
```diff
@@ -34,6 +35,347 @@
+        arkProcessError(ark_mem, ARK_ILL_INPUT, __LINE__, __func__, __FILE__,
+                        "error setting command-line argument: %s %s",
+                        argv[i-1], argv[i]);
+        return ARK_ILL_INPUT;
```
It seems this function always returns `ARK_ILL_INPUT` on a failure. If the input was ok, but a function like `ARKodeSetInterpolantType` failed for a different reason, e.g., some internal interpolation memory was corrupted, it would be more informative to pass along that error.

Steven-Roberts reviewed (commented):
I like how clean this is! I'm ok with the public API aside from the handling of boolean settings.

gardner48 commented on src/arkode/arkode_io.c:
```diff
@@ -34,6 +35,347 @@
+      continue;
+    }
+
+    if (strcmp(argv[i], "arkode.tolerances") == 0)
```
Maybe this should be `arkode.scalar_tolerances` so we could have `arkode.vector_tolerances`. Setting the vector tolerance would be more complicated though e.g., if using a distributed vector, GPU vector, user-defined vector, etc. Do we plan on supporting vector-valued inputs from the command line?

balos1 replied:
We do not have to support vector-valued inputs in the first pass, but yes, we plan to support vector-valued. The design doc covers non-distributed vectors. For distributed vectors, we will have to think about it more. Initially, I would be inclined to take the first `local_size` vector-value inputs and copy them to all ranks.

balos1 replied:
It would be nice to have separate arguments for absolute and relative tolerances. However, since the tolerance 'set' functions take both as inputs it would be a little complicated to implement. Maybe we could call the tolerance function twice?

gardner48 replied:
> I dont think we need to discuss the distributed aspect here though, we should do it in the design document.

Agreed, we'll need to work through this for GPU and ManyVectors also. I think we should be able to leverage the buffer pack and unpack functions to make these work.

> Maybe we could call the tolerance set function twice?

I think this would be a little tricky since we need to save the values for the later calls

drreynolds commented:
I think I've finished revising the previous implementation according to the initial feedback.  It would be great if folks who looked at this previously would give it another look.  Once you're happy with the structure, I'll get started on the ARKODE stepper-specific functions, other integrators, linear solvers, nonlinear solvers, etc.

Steven-Roberts commented on src/arkode/arkode_io.c:
```diff
@@ -34,6 +35,360 @@
   ARKODE optional input functions
   ===============================================================*/
 
+/*---------------------------------------------------------------
```
[Optional suggestion]: Move the new content in `arkode_io.c` to a separate file like `arkode_cli.c` since this file is already close to 4k lines

Steven-Roberts commented on src/arkode/arkode_io.c:
```diff
@@ -34,6 +35,360 @@
+     {"small_num_efails", ARKodeSetSmallNumEFails},
+     {"max_err_test_fails", ARKodeSetMaxErrTestFails},
+     {"max_conv_fails", ARKodeSetMaxConvFails}};
+  int num_int_keys = 15;
```
```suggestion
  static const int num_int_keys = sizeof(int_pairs) / sizeof(*int_pairs);
```

Steven-Roberts commented on src/arkode/arkode_io.c:
```diff
@@ -34,6 +35,360 @@
+     {"max_first_growth", ARKodeSetMaxFirstGrowth},
+     {"max_efail_growth", ARKodeSetMaxEFailGrowth},
+     {"max_cfail_growth", ARKodeSetMaxCFailGrowth}};
+  int num_real_keys = 18;
```
```suggestion
  static const int num_real_keys = sizeof(real_pairs) / sizeof(*real_pairs);
```

Steven-Roberts commented on src/arkode/arkode_io.c:
```diff
@@ -34,6 +35,360 @@
+     {"clear_stop_time", ARKodeClearStopTime},
+     {"no_inactive_root_warn", ARKodeSetNoInactiveRootWarn},
+     {"reset_accumulated_error", ARKodeResetAccumulatedError}};
+  int num_action_keys = 4;
```
```suggestion
  static const int num_action_keys = sizeof(action_pairs) / sizeof(*action_pairs);
```

Steven-Roberts commented on src/arkode/arkode_io.c:
```diff
@@ -34,6 +35,360 @@
+
+  static struct arkKeyLongPair long_pairs[] = {
+    {"max_num_steps", ARKodeSetMaxNumSteps}};
+  int num_long_keys = 1;
```
```suggestion
  static const int num_long_keys = sizeof(long_pairs) / sizeof(*long_pairs);
```

Steven-Roberts commented on src/arkode/arkode_io.c:
```diff
@@ -34,6 +35,360 @@
+    sunbooleantype arg_used = SUNFALSE;
+
+    /* skip command-line arguments that do not begin with "arkode." */
+    static const char* prefix = "arkode.";
```
Just repeating my message on Slack for posterity

What about this format: `<arkid>.<option>` where `arkid` defaults to `arkode.` if `NULL`. I think this has two benefits:
- Users can better control nesting and prefix of option
- If a user wants to swap integrators without rewriting each option, they can use an empty or consistent prefix among the command line parsing functions.

Steven-Roberts commented on include/sundials/priv/sundials_cli.h:
```diff
@@ -0,0 +1,148 @@
+/* -----------------------------------------------------------------
```
Can this be moved to `src/sundials/sundials_cli.h` to limit scope? I don't think any of the declarations are used in `include`.

gardner48 replied:
`include/sundials/priv` is installed because some header files include function implementations so the "private" error handling functions and macros need to be installed so they're accessible in these functions. If these implementations were made internal then `include/sundials/priv` could be removed.

Steven-Roberts commented on include/arkode/arkode.h:
```diff
@@ -214,6 +214,10 @@ typedef enum
  * -------------------------- */
 
+/* Command-line control over ARKODE options */
+SUNDIALS_EXPORT int ARKodeSetFromCommandLine(void* arkode_mem, const char* arkid,
+                                             int argc, char* argv[]);
```
```suggestion
SUNDIALS_EXPORT int ARKodeSetFromCommandLine(void* arkode_mem, const char* arkid,
                                             int argc, const char* argv[]);
```
I'm not sure if I suggested this already... Functions with this arguments will also need to be updated.

drreynolds replied:
You did suggest this previously, but I could not get it to work.  If I recall the error messages correctly, the problem is that this it is essentially a `char**` so the compiler cannot ensure that functions will not interfere with its internal memory.  I ended up having to propagate the `const` all the way to the `main` routine for it to compile, and I believe that is overly restrictive for user code (who may want to remove command-line options after they are parsed).

Steven-Roberts commented on src/arkode/arkode_cli.c:
```diff
@@ -0,0 +1,312 @@
+      continue;
+    }
+
+    if (strcmp(argv[i] + offset, "writeparameters") == 0)
```
Is the convention snake case? Either way, I like that this option is included.
```suggestion
    if (strcmp(argv[i] + offset, "write_parameters") == 0)
```

Steven-Roberts commented on src/sunadaptcontroller/imexgus/sunadaptcontroller_imexgus.c:
```diff
@@ -94,6 +107,92 @@ SUNAdaptController SUNAdaptController_ImExGus(SUNContext sunctx)
+    {
+      i += 1;
+      sunrealtype rarg = atof(argv[i]);
+      retval           = SUNAdaptController_SetErrorBias_ImExGus(C, rarg);
```
Since bias can be set to a generic `SUNAdaptController` should we parse the "error_bias" argument in the superclass rather than duplicating it in each subclass?

drreynolds replied:
The reason I put it here is so that if a user wants to set both `error_bias` and `params`, those would use the same prefix.  If `error_bias` were moved to the superclass, then it seems to me that the user would need to differentiate the prefixes even though they're using the same controller, e.g.,
```
  a.out sunadaptcontroller.error_bias 1.2 sunadaptcontroller_imexgus.params 1.0 1.0 1.0 1.0
```

I'd rather have a small amount of duplicated code between the three adaptivity controllers than cause that level of confusion for users.

drreynolds replied:
I'm updating everything (SUNAdaptControllers, SUNLinearSolvers, SUNNonlinearSolvers) so that each shares a common prefix (e.g., "sunadaptcontroller" or "sunlinearsolver").  However, the new "feature" wherein we will support either command-line or file-based arguments using the same class function necessitates that the actual processing be handled in each implementation and not directly in the base class.  While it will be somewhat redundant, it will certainly clean up the API and allow us to be more "forward thinking" with respect to file-based inputs.

drreynolds replied:
I am concerned with how the branching (file-based input vs command-line input) would occur at two separate levels, given the single `SetOptions` points of entry for both the base class and the implementations.  If we want the base class to handle some keys before calling the implementations, then it would need to first branch between file-based vs command-line input, call its private function for file-based input, call its private function for command-line input, and then call the implementation's `SetOptions` function pointer to repeat the process.  Since both the base class and implementations will share the same prefix (here, `sunadaptcontroller`), then I am concerned that doing (a) base class file-based, then (b) base class command-line, then (c) implementation file-based, then (d) implementation command-line, could open things up for odd behavior if the keys were not sufficiently distinct.

Steven-Roberts commented on src/sunadaptcontroller/imexgus/sunadaptcontroller_imexgus.c:
```diff
@@ -94,6 +107,92 @@ SUNAdaptController SUNAdaptController_ImExGus(SUNContext sunctx)
+     processes if requested. */
+  if (write_parameters)
+  {
+    retval = SUNAdaptController_Write_ImExGus(C, stdout);
```
Same issue about moving this parsing up to superclass

drreynolds replied:
Same response as for `error_bias`

gardner48 replied:
Continuing the discussion in [this](https://github.com/LLNL/sundials/pull/662#discussion_r2072196366) thread

Steven-Roberts commented on src/sunadaptcontroller/soderlind/sunadaptcontroller_soderlind.c:
```diff
@@ -103,6 +117,187 @@ SUNAdaptController SUNAdaptController_Soderlind(SUNContext sunctx)
+    /* if Cid is supplied, skip command-line arguments that do not begin with Cid;
+       else, skip command-line arguments that do not begin with "sunadaptcontroller." */
+    size_t SOffset, PIDOffset, PIOffset, IOffset, ExpGusOffset, ImpGusOffset;
+    const char* SPrefix   = (strlen(Cid) > 0) ? Cid
+                                              : "sunadaptcontroller_soderlind.";
+    const char* PIDPrefix = (strlen(Cid) > 0) ? Cid : "sunadaptcontroller_pid.";
+    const char* PIPrefix  = (strlen(Cid) > 0) ? Cid : "sunadaptcontroller_pi.";
+    const char* IPrefix   = (strlen(Cid) > 0) ? Cid : "sunadaptcontroller_i.";
+    const char* ExpGusPrefix = (strlen(Cid) > 0) ? Cid
+                                                 : "sunadaptcontroller_expgus.";
+    const char* ImpGusPrefix = (strlen(Cid) > 0) ? Cid
+                                                 : "sunadaptcontroller_impgus.";
```
I think a couple more controllers were added in 7.3 so this would need to be updated. The bigger issues are this function is difficult to maintain and it breaks what I think is a convention of associating one prefix per class. I'm struggling to think of a good solution, but moving bias, defaults, and write parameter parsing up to superclass will help.

drreynolds replied:
I disagree.  The new controllers that were added in 7.3 have fixed coefficients, and thus they do not even provide `SetParams` routines in the first place.  This piece had to be complicated to support our older controller API, where the controller "type" merely indicated the specific structure of the controller (one vs two vs three (h,dsm) history values, etc), but all exponents were user modifiable.

Since the new controllers have no `SetParams` routines, then their other parameters (e.g., bias) can already be specified via the `Cid` value `sunadaptcontroller_soderlind`.  I added this note to the documentation for each of those utility constructors.

Steven-Roberts reviewed (changes requested):
Not necessary for this PR, but some sort of help functionality would be very helpful for identifying available options.

gardner48 commented on src/sunadaptcontroller/imexgus/sunadaptcontroller_imexgus.c:
```diff
@@ -94,6 +107,92 @@ SUNAdaptController SUNAdaptController_ImExGus(SUNContext sunctx)
+    /* control over SetParams function */
+    if (strcmp(argv[i] + offset, "params") == 0)
+    {
+      i += 1;
+      sunrealtype rarg1 = atof(argv[i]);
+      i += 1;
+      sunrealtype rarg2 = atof(argv[i]);
+      i += 1;
+      sunrealtype rarg3 = atof(argv[i]);
+      i += 1;
+      sunrealtype rarg4 = atof(argv[i]);
```
The function is `sunrealtype SUNStrToReal(const char* str)` in `sundials_math.h|c`

balos1 commented on src/sundials/sundials_cli.c:
```diff
@@ -0,0 +1,211 @@
+    (*i) += 1;
+    int iarg = atoi(argv[*i]);
+    (*i) += 1;
+    sunrealtype rarg = atof(argv[*i]);
```
```suggestion
    sunrealtype rarg = SUNStrToReal(argv[*i]);
```

balos1 commented on src/sundials/sundials_cli.c:
```diff
@@ -0,0 +1,211 @@
+    (*i) += 1;
+    int iarg = atoi(argv[*i]);
+    (*i) += 1;
+    sunrealtype rarg1 = atof(argv[*i]);
```
```suggestion
    sunrealtype rarg1 = SUNStrToReal(argv[*i]);
```

balos1 commented on src/sundials/sundials_cli.c:
```diff
@@ -0,0 +1,211 @@
+    (*i) += 1;
+    sunrealtype rarg1 = atof(argv[*i]);
+    (*i) += 1;
+    sunrealtype rarg2 = atof(argv[*i]);
```
```suggestion
    sunrealtype rarg2 = SUNStrToReal(argv[*i]);
```

balos1 commented on src/sundials/sundials_cli.c:
```diff
@@ -0,0 +1,211 @@
+  if (strcmp(argv[*i] + offset, argtest) == 0)
+  {
+    (*i) += 1;
+    sunrealtype rarg = atof(argv[*i]);
```
```suggestion
    sunrealtype rarg = SUNStrToReal(argv[*i]);
```

balos1 commented on src/sundials/sundials_cli.c:
```diff
@@ -0,0 +1,211 @@
+  if (strcmp(argv[*i] + offset, argtest) == 0)
+  {
+    (*i) += 1;
+    sunrealtype rarg1 = atof(argv[*i]);
```
```suggestion
    sunrealtype rarg1 = SUNStrToReal(argv[*i]);
```

balos1 commented on src/sundials/sundials_cli.c:
```diff
@@ -0,0 +1,211 @@
+    (*i) += 1;
+    sunrealtype rarg1 = atof(argv[*i]);
+    (*i) += 1;
+    sunrealtype rarg2 = atof(argv[*i]);
```
```suggestion
    sunrealtype rarg2 = SUNStrToReal(argv[*i]);
```

balos1 reviewed (changes requested):
The comments I have left are pretty minor. 

My two biggest concerns are 1) that developers will have to remember that the *_cli.c files need to be updated when a set function is added, 2) its hard to test that everything is wired up correctly. 

Unfortunately, I don't have a great way to address either of these concerns, but one thing that may help is a developer documentation section on "Setting SUNDIALS Options through the Command Line" where we somewhat briefly describe how this all works. Then, we can also add a note to the [developer checklist](https://sundials.readthedocs.io/en/latest/developers/getting_started/Checklist.html) about having to update the CLI files when a set function is added.

drreynolds commented on src/arkode/arkode_arkstep_io.c:
```diff
@@ -628,6 +629,63 @@ int ARKStepGetTimestepperStats(void* arkode_mem, long int* expsteps,
+
+  /* check all "twochar" command-line options */
+  int j, retval;
+  for (j = 0; j < num_twochar_keys; j++)
+  {
+    retval = sunCheckAndSetTwoCharArg((void*)ark_mem, i, argv, offset,
+                                      twochar_pairs[j].key,
+                                      twochar_pairs[j].set, arg_used);
+    if (retval != ARK_SUCCESS)
+    {
+      arkProcessError(ark_mem, retval, __LINE__, __func__, __FILE__,
+                      "error setting command-line argument: %s",
+                      twochar_pairs[j].key);
+      return retval;
+    }
+    if (*arg_used) { return ARK_SUCCESS; }
+  }
```
Given that each integrator still has its own mechanism for error handling, I did not create a SUNDIALS-wide helper for this.  I'll consider whether a SUNDIALS-wide helper function could just use SUNDIALS-based error handling, and then each intgrator could just test whether the overall helper function failed before calling the integrator-specific error hander.

drreynolds commented on src/sunlinsol/klu/sunlinsol_klu.c:
```diff
@@ -168,6 +181,47 @@ SUNErrCode SUNLinSol_KLUReInit(SUNLinearSolver S, SUNMatrix A, sunindextype nnz,
+    }
+
+    /* control over SetOrdering function */
+    if (strcmp(argv[i] + offset, "ordering") == 0)
```
I don't see how that would be possible.

The issue is that `sunCheckAndSetIntArg` expects that the "Set" routine provided must have type `int (*sunIntSetFn)(void*, int)`.  However, the first argument to each linear solver module set routine has type `SUNLinearSolver`.  This same pattern is present for our other object-oriented modules as well.  It would be a huge API change to convert all our object-oriented module set routines to require `void*` input (and then cast the argument to the appropriate type), just so that the command-line processing function could call the shared routine.

Steven-Roberts replied:
Ah, I didn't realize the first argument type wasn't `void*`.

drreynolds commented:
I believe I have addressed all PR comments. Please re-review as needed.

gardner48 commented on src/sunlinsol/klu/sunlinsol_klu.c:
```diff
@@ -168,6 +181,47 @@ SUNErrCode SUNLinSol_KLUReInit(SUNLinearSolver S, SUNMatrix A, sunindextype nnz,
+    if (LSid != NULL)
+    {
+      if (strncmp(argv[idx], LSid, strlen(LSid)) != 0) { continue; }
+      offset = strlen(LSid) + 1;
```
See related discussion [here](https://github.com/LLNL/sundials/pull/662#discussion_r2197782093) about supporting an empty string as the prefix.

gardner48 commented on src/arkode/arkode_sprkstep_io.c:
```diff
@@ -194,6 +195,48 @@ int SPRKStepGetNumRhsEvals(void* arkode_mem, long int* nf1, long int* nf2)
+  static const int num_char_keys = sizeof(char_pairs) / sizeof(*char_pairs);
+
+  static const struct sunKeyIntPair int_pairs[] = {
+    {"use_compensated_sums", SPRKStepSetUseCompensatedSums}};
```
I don't think I noted this elsewhere but `ARKodeSetUseCompensatedSums` need to be added to `arkode_cli.c`

gardner48 commented on src/sunadaptcontroller/soderlind/sunadaptcontroller_soderlind.c:
```diff
@@ -103,6 +117,184 @@ SUNAdaptController SUNAdaptController_Soderlind(SUNContext sunctx)
+  {
+    /* if Cid is supplied, skip command-line arguments that do not begin with Cid;
+       else, skip command-line arguments that do not begin with "sunadaptcontroller." */
+    size_t SOffset, PIDOffset, PIOffset, IOffset, ExpGusOffset, ImpGusOffset;
```
See related discussion in the conversation thread [here](https://github.com/LLNL/sundials/pull/662#discussion_r2072196366)

gardner48 commented:
@drreynolds I jumped around the files a little in reviewing, so you'll probably want to read through the less trivial comments before diving into updates.

drreynolds commented on doc/arkode/guide/source/Usage/User_callable.rst:
```diff
@@ -906,6 +912,39 @@ Set the checkpointing step index (for adjoint)     :c:func:`ARKodeSetAdjointChec
 
 
 
+.. c:function:: int ARKodeSetFromCommandLine(void* arkode_mem, const char* arkid, int argc, char* argv[])
```
Yes, the goal is to eventually support input files.  I'll add the extra layer and update the current routines, but this will take some time since there are now many of these interspersed throughout the repository.

drreynolds replied:
In thinking about this further, we'd need to give precedence to either command-line or file input options with this approach.  Perhaps if `file_name` is non-NULL then those are processed first, followed by command-line options?  For now we would leave the non-NULL `file_name` conditional as a "To-Do" item, and note in the documentation that it is not yet implemented.

gardner48 replied:
Yes, that's what I'm thinking and `SetOptions` could just be a wrapper function something like
```
int ARKodeSetOptions(void* arkode_mem, const char* arkid, const char* file_name, int argc, char* argv[])
{
  if (file_name != NULL && strlen(empty_string) != 0)
  {
    int retval = arkSetOptionsFromFile(arkode_mem, arkid, file_name);
    if (retval != ARK_SUCCESS) { return retval; }
  }

  if (argc > 0 && argv != NULL)
  {
    int retval = arkSetOptionsFromCommandLine(arkode_mem, arkid, argc, argv);
    if (retval != ARK_SUCCESS) { return retval; }
  }

  return ARK _SUCCESS;
}
```
For now, we would not call `arkSetOptionsFromFile` and instead return an error that this is not supported yet.

drreynolds replied:
What is the role of the `&& strlen(empty_string) != 0` here?

gardner48 commented on doc/shared/sunadaptcontroller/SUNAdaptController_Description.rst:
```diff
@@ -261,13 +267,47 @@ note these requirements below. Additionally, we note the behavior of the base SU
+   .. note::
+
+      This routine will be called by :c:func:`SUNAdaptController_SetFromCommandLine`
+      when using the command-line option "Cid.defaults".
```
I think I made this comment before getting into the source. I see now that the inputs are processed in the order they are provided (we should note this in the user guide if we don't already) so the user will just need to consider any option order dependencies when providing the inputs. 

Usually we don't require a particular order, so maybe this is generally not an issue, but it's something we'll need to watch out for when adding new set functions and, to make things easier for the user, we should try to avoid adding an order dependence when possible.

drreynolds commented on examples/arkode/CXX_serial/ark_heat2D_lsrk.cpp:
No, I believe that I'd tested some options with this file (which required me to remove the extraneous error check whether too-many options were supplied), but once it passed I removed the extra code.

drreynolds commented:
I believe I've addressed all PR comments *except* for moving command-line processing to base classes from implementations.  As I noted in my responses, I have concerns with making that move, since I worry that it decreases robustness for the sake of saving a few lines of code from implementations (but adding around the same number of lines elsewhere). 

If others feel differently then I recommend we debate that later in a subsequent PR.  I see no reason such a minor implementation detail should hold up this release, since even if we decide to eventually re-engineer the internals to make the move, it will not change the functionality or API.

I'm still updating answer files due to changes in what is printed in some examples, but otherwise this PR should be good for re-review.
